test/Shared/TestBase.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if !NET7_0_OR_GREATER
+using System;
+using System.IO;
+#endif
+using Microsoft.TemplateEngine.TestHelper;
+
+namespace Microsoft.TemplateEngine.Tests
+{
+    /// <summary>
+    /// The class contains the utils for unit and integration tests.
+    /// </summary>
+    public abstract class TestBase
+    {
+        internal static string CodeBaseRoot { get; } = GetCodeBaseRoot();
+
+        internal static string ShippingPackagesLocation
+        {
+            get
+            {
+#if DEBUG
+                string configuration = "Debug";
+#elif RELEASE
+                string configuration = "Release";
+#else
+                throw new NotSupportedException("The configuration is not supported");
+#endif
+
+                string packagesLocation = Path.Combine(CodeBaseRoot, "artifacts", "packages", configuration, "Shipping");
+
+                if (!Directory.Exists(packagesLocation))
+                {
+                    throw new Exception($"{packagesLocation} does not exist");
+                }
+                return Path.GetFullPath(packagesLocation);
+            }
+        }
+
+        internal static string TestTemplatesLocation { get; } = Path.Combine(CodeBaseRoot, "test", "Microsoft.TemplateEngine.TestTemplates", "test_templates");
+
+        internal static string TestTemplatePackagesLocation { get; } = Path.Combine(CodeBaseRoot, "test", "Microsoft.TemplateEngine.TestTemplates", "nupkg_templates");
+
+        internal static string TestPackageProjectPath { get; } = Path.Combine(CodeBaseRoot, "test", "Microsoft.TemplateEngine.TestTemplates", "Microsoft.TemplateEngine.TestTemplates.csproj");
+
+        internal static string PackTestTemplatesNuGetPackage(PackageManager packageManager)
+        {
+            return packageManager.PackNuGetPackage(TestPackageProjectPath);
+        }
+
+        internal static string GetTestTemplateLocation(string templateName)
+        {
+            string templateLocation = Path.Combine(TestTemplatesLocation, templateName);
+
+            if (!Directory.Exists(templateLocation))
+            {
+                throw new Exception($"{templateLocation} does not exist");
+            }
+            return Path.GetFullPath(templateLocation);
+        }
+
+        private static string GetCodeBaseRoot()
```

Comment:
This still seems to be pretty repo-specific.
Though there is no much better way - either we'd need to deploy the templates to output (not a good idea), or we need to try to guess the repo root (likely preferred) as is done here. But in case of guessing we should try to fail early and clearly by some validation mechanism. e.g. `if (!File.Exists(Path.Combine(codeBaseRoot, "LICENSE.TXT"))) { /* Error out */}`.

Reply:
Or better - https://github.com/dotnet/templating/pull/5118/files#diff-ece9a306b3c54b1e8b78cdebdc6fc4683079669976096a4c88e20bc1e5e73cb8L65

Reply:
This `TestBase` is very repo specific, it is not shipped with any of the assembly. I will add more sanity checks here so we don't continue if the structure significantly changes.